packages/fast-tooling-react/src/form/utilities/navigation.ts

```diff
@@ -0,0 +1,381 @@
+}
+
+export interface Component {
+    dataLocation: string;
+    schema: unknown;
+}
+
+export interface NavigationConfig {
+    dataLocation: string;
+    data: unknown | void;
+    schema: unknown;
+    childOptions: FormChildOptionItem[];
+}
+
+class Navigation {
+    private navigationMap: NavigationMap;
+
+    private config: NavigationConfig;
+
+    private dataLocation: string;
+
+    private reactChildrenDataLocations: string[];
+
+    private resolveNormalizeDataLocation: ((dataLocation: string) => string) &
+        MemoizedFunction;
+
+    private resolveDataLocations: ((dataLocation: string) => string[]) & MemoizedFunction;
+
+    constructor(config: NavigationConfig) {
+        this.navigationMap = {};
+        this.config = cloneDeep(config);
+        this.config.childOptions = getReactDefaultChildren().concat(
+            this.config.childOptions
+        );
+        this.resolveNormalizeDataLocation = memoize(this.normalizeDataLocation);
+        this.resolveDataLocations = memoize(this.getDataLocationArray);
+        this.updateChildrenLocations();
+        this.dataLocation = this.resolveNormalizeDataLocation(config.dataLocation);
+
+        this.setItemToNavigationMap(this.dataLocation);
+    }
+
+    /**
+     * Update the location, if not already part of the mapping
+     */
+    public updateDataLocation(
+        dataLocation: string,
+        callback: (updatedNav: NavigationItem[]) => void
+    ): void {
+        this.dataLocation = this.resolveNormalizeDataLocation(dataLocation);
+
+        if (
+            typeof get(this.navigationMap[dataLocation], "schemaLocation") === "undefined"
+        ) {
+            this.setItemToNavigationMap(this.dataLocation);
+        }
+
+        callback(this.get());
+    }
+
+    /**
+     * Update data at the current data location
+     */
+    public updateData(
+        data: unknown,
+        callback: (updatedNav: NavigationItem[]) => void
+    ): void {
+        this.config.data = cloneDeep(data);
+        this.updateChildrenLocations(); // children may have been added/removed
+
+        Object.keys(this.navigationMap).forEach((navigationMapKey: string) => {
+            if (typeof this.navigationMap[navigationMapKey].data !== "undefined") {
+                if (navigationMapKey === "") {
+                    this.navigationMap[""].data = this.config.data;
+                } else {
+                    this.navigationMap[navigationMapKey].data = get(
+                        this.config.data,
+                        this.navigationMap[navigationMapKey].dataLocation
+                    );
+                }
+            }
+        });
+
+        callback(this.get());
+    }
+
+    /**
+     * Gets an array of navigation items
+     */
+    public get(): NavigationItem[] {
+        const navigationItems: NavigationItem[] = this.resolveDataLocations(
+            this.dataLocation
+        ).map(
+            (dataLocation: string): NavigationItem => {
```

Comment:
You're looking this array twice with different map functions - you should be able to combine this into one loop I think

Reply:
I actually need to do it twice because until its converted to an array I don't know how to resolve inherited default values, I've simplified it though!